acceptance/container.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path/filepath"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/archive"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+var containerClient *client.Client
+
+const containerImage = "local.build-trusted-artifacts:acceptance"
+
+func init() {
+	var err error
+	containerClient, err = client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		panic(fmt.Sprintf("creating new client: %s", err))
+	}
+}
+
+func runContainer(ctx context.Context, cmd []string, binds []string) error {
+	cont, err := containerClient.ContainerCreate(
+		ctx,
+		&container.Config{
+			Image: containerImage,
+			Tty:   true, // Prevent leading metadata characters in the container logs... weird
+			Cmd:   cmd,
+		},
+		&container.HostConfig{
+			Binds: binds,
+		},
+		&network.NetworkingConfig{},
+		&ocispec.Platform{},
+		"", // Let docker pick a name.
+	)
+	if err != nil {
+		return fmt.Errorf("creating container: %w", err)
+	}
+
+	if err := containerClient.ContainerStart(ctx, cont.ID, container.StartOptions{}); err != nil {
+		return fmt.Errorf("starting container %s: %w", cont.ID, err)
+	}
+
+	if err := waitForContainer(ctx, cont.ID); err != nil {
+		return fmt.Errorf("waiting for container %s: %w", cont.ID, err)
+	}
+
+	if err := containerClient.ContainerRemove(ctx, cont.ID, container.RemoveOptions{}); err != nil {
```

Comment:
Should this be in a defer right after `ContainerCreate`, otherwise could lead to dangling containers in case of errors in `ContainerStart` or `ContainerWait`. This probably should also include a call to `ContainerStop`?
Something like this on line 48:

```
	defer func() {
		_ := containerClient.ContainerStop(ctx, cont.ID, container.StopOptions{})
		_ := containerClient.ContainerRemove(ctx, cont.ID, container.RemoveOptions{})
	}()

Reply:
Yes, good call. I did not add a call to `ContainerStop`. Instead, it's now calling `ContainerRemove` with `Force: true`.